### PR TITLE
Feature/drizshko/ngraph constant mem reuse

### PR DESCRIPTION
**Pull request by Drizshko**
This is continued  work on proposal for mmap allocator. I think it will be better to break this task into 2 parts
1. Implement memory re-use for Constant node. After that it will be possible to use mmap allocator as a part of user code (via CBlob)
2. Implement internal mmap allocator.

### CODE REVIEW

File: inference-engine/include/cpp/ie_cnn_network.h
Comment (ilyachur) [inline inference-engine/include/cpp/ie_cnn_network.h:195]: Why do we need to extend the public API?
At the current moment we read model and weights at the same time.

Comment (Drizshko) [reply inference-engine/include/cpp/ie_cnn_network.h:195]: Right, this one should be private or protected as not intended to be used thru public APIs. I'll fix that.

Comment (Drizshko) [reply inference-engine/include/cpp/ie_cnn_network.h]: Fixed in the latest commit.

File: inference-engine/include/cpp/ie_cnn_network.h
Comment (ilyachur) [inline inference-engine/include/cpp/ie_cnn_network.h:250]: Why do we need this method?

Comment (Drizshko) [reply inference-engine/include/cpp/ie_cnn_network.h:250]: Need somehow to save a pointer to blob. Lifetime of blob should be the same as for CNNNetwork. 
Idea here  is to read model data   to blob, provide a pointer  to Constant nodes and make them wrap already allocated memory instead of creating new buffers. 
I can do the same as for Blob Proxy and make blob consumers keep pointers, but in this case i would need to expose InferenceEngine types and interfaces to nGraph. Personally i don't like this idea as it will make nGraph dependent on IE, so i;m trying to keep them saparate.

Comment (ilya-lavrenov) [reply inference-engine/include/cpp/ie_cnn_network.h:250]: BTW, I don't see where this method is used. We should not touch public API for internal needs..

Comment (ilyachur) [reply inference-engine/include/cpp/ie_cnn_network.h:251]: @Drizshko could you provide example or test which will show why we need this method and friend class V10Parser here?

Comment (Drizshko) [reply inference-engine/include/cpp/ie_cnn_network.h]: @ilyachur @ilya-lavrenov i did change implementation, now nGraph buffer itself will store smart pointer to Blob. 
Did this without linking nGraph to IE, by using template parameter.
Please take a look.

File: ngraph/core/include/ngraph/runtime/aligned_buffer.hpp
Comment (ilyachur) [inline ngraph/core/include/ngraph/runtime/aligned_buffer.hpp:80]: Maybe we can just move some fields to protected section instead of make a friend class.

**Review by ilyachur — COMMENTED** [review]
It looks better for me than the variant with introducing new methods to public API 😄 
Also as I remember the original idea was to use mmap for weights.
As I understand at the current moment we don't use it in this PR and don't have any examples how user can do it, am I right?

**Comment by Drizshko** [comment]
> It looks better for me than the variant with introducing new methods to public API 😄
> Also as I remember the original idea was to use mmap for weights.
> As I understand at the current moment we don't use it in this PR and don't have any examples how user can do it, am I right?


I did refactoring, did remove friend class, changed access to protected and did move new class declaration to new header.

This is first step to add mmap allocator, it has 2 parts  - make Constant to reuse memory (what i'm doing in this PR) and next to create MmapAllocator - next step. It will be easier to review and implement.

**Comment by ilyachur** [comment]
Could you fix CI?

File: inference-engine/tests/ie_test_utils/unit_test_utils/mocks/mock_not_empty_icnn_network.hpp
Comment (ilya-lavrenov) [inline inference-engine/tests/ie_test_utils/unit_test_utils/mocks/mock_not_empty_icnn_network.hpp:80]: remove?

File: ngraph/core/include/ngraph/runtime/shared_buffer.hpp
Comment (ilya-lavrenov) [inline ngraph/core/include/ngraph/runtime/shared_buffer.hpp:21]: `"ngraph/runtime/aligned_buffer.hpp"`

**Review by ilya-lavrenov — APPROVED** [review]

**Comment by vurusovs** [comment]
@Drizshko several memory tests are failed on SSD300 on CPU:
Reference RSS memory consumption: 307990 KB.
Current RSS memory consumption: 334632 KB.

Please, contact me or @asomsiko if any help is required

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_normalize_node.cpp
Comment (ilyachur) [inline inference-engine/src/mkldnn_plugin/nodes/mkldnn_normalize_node.cpp:709]: Why do you need to copy this blob?

File: inference-engine/src/readers/ir_reader_v7/ie_ir_parser.cpp
Comment (ilyachur) [inline inference-engine/src/readers/ir_reader_v7/ie_ir_parser.cpp:62]: I think you should pass empty `weightsPtr` without this fake creation.

Comment (Drizshko) [reply inference-engine/src/readers/ir_reader_v7/ie_ir_parser.cpp:62]: in another case test without weights crashes with exception while  trying to access blob. In previous implementation it passes empty stream with 0 length, so i did  the same - zero size blob.

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_normalize_node.cpp
Comment (Drizshko) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_normalize_node.cpp:709]: I just  tried to avoid references from  MKLDNN node to CNNNetwork. On MKLDNNNode cleanup() we releasing reference to CNNNetLayer to free-up CNNNetwork, but this one still keeps it alive.

File: inference-engine/src/readers/ir_reader_v7/ie_ir_parser.cpp
Comment (ilyachur) [reply inference-engine/src/readers/ir_reader_v7/ie_ir_parser.cpp:62]: Ok got it.

**Review by ilyachur — APPROVED** [review]

**Comment by ilyachur** [comment]
Looks like CI hovered. Could you resolve conflicts and re-run CI?

**Comment by ilya-lavrenov** [comment]
@Drizshko @ilyachur looks like CI is green. Can we merge? If yes, suggest to wait 2 days once release branch is created and merge to master then

**Comment by ilyachur** [comment]
> @Drizshko @ilyachur looks like CI is green. Can we merge? If yes, suggest to wait 2 days once release branch is created and merge to master then

It is why I didn't merge it yesterday. Let's wait for release branch creation and after that merge this PR to master.

**Comment by Drizshko** [comment]
Merged with latest master, @ilyachur can we merge now?

**Comment by ilyachur** [comment]
> Merged with latest master, @ilyachur can we merge now?

I would like to merge this PR after the creation of release branch.

**Comment by ilyachur** [comment]
@Drizshko Could you please rebase this PR to the latest master?

**Comment by Drizshko** [comment]
Done

File: inference-engine/src/legacy_api/CMakeLists.txt
Comment (ilyachur) [inline inference-engine/src/legacy_api/CMakeLists.txt:25]: Why do we need this dummy file?

**Pull request merged**
